core/src/main/java/org/apache/iceberg/io/RollingFileWriter.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A rolling writer capable of splitting incoming data/deletes into multiple files within one spec/partition.
+ */
+abstract class RollingFileWriter<T, W extends FileWriter<T, R>, R> implements FileWriter<T, R> {
+  private static final int ROWS_DIVISOR = 1000;
+
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final FileFormat fileFormat;
+  private final long targetFileSizeInBytes;
+  private final PartitionSpec spec;
+  private final StructLike partition;
+
+  private EncryptedOutputFile currentFile = null;
+  private W currentWriter = null;
+  private long currentRows = 0;
+
+  private boolean closed = false;
+
+  protected RollingFileWriter(OutputFileFactory fileFactory, FileIO io, FileFormat fileFormat,
+                              long targetFileSizeInBytes, PartitionSpec spec, StructLike partition) {
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.fileFormat = fileFormat;
+    this.targetFileSizeInBytes = targetFileSizeInBytes;
+    this.spec = spec;
+    this.partition = partition;
+  }
+
+  protected abstract W newWriter(EncryptedOutputFile file);
+
+  protected abstract long length(W writer);
+
+  protected abstract void addResult(R result);
+
+  protected abstract R aggregatedResult();
+
+  protected PartitionSpec spec() {
+    return spec;
+  }
+
+  protected StructLike partition() {
+    return partition;
+  }
+
+  @Override
+  public void write(T row) throws IOException {
+    currentWriter.write(row);
+    currentRows++;
+
+    if (shouldRollToNewFile()) {
+      closeCurrent();
+      openCurrent();
+    }
+  }
+
+  public CharSequence currentPath() {
+    Preconditions.checkNotNull(currentFile, "The currentFile shouldn't be null");
+    return currentFile.encryptingOutputFile().location();
+  }
+
+  public long currentRows() {
+    return currentRows;
+  }
+
+  protected void openCurrent() {
+    if (partition == null) {
+      this.currentFile = fileFactory.newOutputFile();
+    } else {
+      this.currentFile = fileFactory.newOutputFile(partition);
+    }
+    this.currentWriter = newWriter(currentFile);
+    this.currentRows = 0;
+  }
+
+  private boolean shouldRollToNewFile() {
+    // TODO: ORC file now not support target file size before closed
+    return !fileFormat.equals(FileFormat.ORC) &&
+        currentRows % ROWS_DIVISOR == 0 && length(currentWriter) >= targetFileSizeInBytes;
+  }
+
+  private void closeCurrent() throws IOException {
+    if (currentWriter != null) {
+      currentWriter.close();
+
+      if (currentRows == 0L) {
+        io.deleteFile(currentFile.encryptingOutputFile());
+      } else {
+        R result = currentWriter.result();
+        addResult(result);
```

Comment:
Minor: do we need `result` or can we just call `addResult(currentWriter.result())`?